### PR TITLE
Winkd fixes

### PR DESCRIPTION
**Pull request by GustavoLCR**
# Please don't squash
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

4c0bac5c7b80a7c1ac14eec5098a4cd616b2a449 Get memory maps, it's still missing permissions (blocked by `RzAnalysis`/`PDB` parser not supporting bitfields yet), and the file name info
585fec40b1e0635607fb1f86c58b7a9987f523ac Fixes reading big sections of memory at once
9717e455667812c5e09e94e1610a91bdb7124984 Fix to completely write the register values
1509d968937d2a8a93e023e3476f4cc6f79998a6 Is the only solution for now to allow breakpoints in early initialization when there are no valid VADs yet
035c9873573b615d742da30d9485fbcccbc2d0b0 Fixes some desyncs that caused rizin to confuse register read with memory read packets, and other stuff
98f8c7e14846cb038a5689ecd00e6b2a2481dac6 Keep track of current cpu thread/process

**Test plan**

Manually tested

**Closing issues**

Hopefully #2627, I can't repro anymore
cc @PAI5REECHO, can you test and ack the fix?

### CODE REVIEW

File: librz/debug/p/common_windows.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		EXCEPTION_STR(EXCEPTION_INVALID_HANDLE);
+		EXCEPTION_STR(EXCEPTION_NONCONTINUABLE_EXCEPTION);
+		EXCEPTION_STR(EXCEPTION_PRIV_INSTRUCTION);
+		EXCEPTION_STR(EXCEPTION_SINGLE_STEP);
+		EXCEPTION_STR(EXCEPTION_STACK_OVERFLOW);
+		EXCEPTION_STR(STATUS_UNWIND_CONSOLIDATE);
+		EXCEPTION_STR(EXCEPTION_POSSIBLE_DEADLOCK);
+		EXCEPTION_STR(DBG_CONTROL_BREAK);
+		EXCEPTION_STR(CONTROL_C_EXIT);
+	case 0x6ba: return "FILE_DIALOG_EXCEPTION";
```
Comment (XVilka) [inline librz/debug/p/common_windows.c]: Please either use capital letters in hexadecimal numbers, especially to be consistent.

File: librz/debug/p/common_windows.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+void windows_print_exception_event(ut32 pid, ut32 tid, ut32 exception_code, bool second_chance) {
+	bool is_fatal = windows_is_exception_fatal(exception_code);
+	RZ_LOG_INFO("(%" PFMT32u ") %s Exception %04X (%s) in thread %" PFMT32u "\n",
+		pid,
+		is_fatal ? "Fatal" : "Non-fatal",
+		exception_code, get_exception_name(exception_code),
+		tid);
+	if (second_chance) {
+		RZ_LOG_WARN("Second-chance exception!!!\n");
```
Comment (XVilka) [inline librz/debug/p/common_windows.c]: One `!` is enough

File: librz/debug/p/common_windows.c [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	if (is_fatal) {
+		RZ_LOG_INFO("Hint: Use 'dce' continue into exception handler\n");
```
Comment (XVilka) [inline librz/debug/p/common_windows.c]: `Use 'dce' to continue...`

File: librz/debug/p/debug_winkd.c [hunk 4]
```diff
@@ -5,36 +5,72 @@
...
-	return winkd_continue(kdctx);
+	return winkd_continue(kdctx, !sig);
+}
+
+static void get_current_process_and_thread(RzDebug *dbg, ut64 thread_address) {
+	if (!O_(ET_ApcProcess)) {
+		return;
+	}
+	WindThread *thread = winkd_get_thread_at(&kdctx->windctx, thread_address);
+	if (thread) {
```
Comment (XVilka) [inline librz/debug/p/debug_winkd.c]: You can invert this condition and do a quick return if `!thread`

File: librz/debug/p/debug_winkd.c [hunk 5]
```diff
@@ -337,6 +382,56 @@ static RzList *rz_debug_winkd_modules(RzDebug *dbg) {
 	return ret;
 }
 
+#include "native/bt/windows-x64.c"
+
+RzList *rz_debug_winkd_frames(RzDebug *dbg, ut64 at) {
```
Comment (XVilka) [inline librz/debug/p/debug_winkd.c]: Not an API but I recommend adding ownership mark anyway.

File: librz/debug/p/debug_winkd.c [hunk 6]
```diff
@@ -337,6 +382,56 @@ static RzList *rz_debug_winkd_modules(RzDebug *dbg) {
...
+		return NULL;
+	}
+	RzList *ret = NULL;
+	if (kdctx->windctx.is_arm) {
+		// TODO
+	} else {
+		if (kdctx->windctx.is_64bit) {
+			struct context_type_amd64 context = { 0 };
+			backtrace_windows_x64(dbg, &ret, &context);
+		} else {
```
Comment (XVilka) [inline librz/debug/p/debug_winkd.c]: It doesn't support x86 32 bytes?
I wonder also, if it makes sense to print warning that it's not implemented, also for ARM.

File: subprojects/rzwinkd/winkd.c [hunk 7]
```diff
@@ -369,54 +387,103 @@ int winkd_wait_packet(KdCtx *ctx, const uint32_t type, kd_packet_t **p) {
...
+	if (left) {
+		winkd_walk_vadtree(ctx, left, self, out);
+	}
+	if (right) {
+		winkd_walk_vadtree(ctx, right, self, out);
+	}
 }
 
-#endif
+RzList *winkd_list_maps(WindCtx *ctx) {
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: Same, I recommend to add ownership mark

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good overall

File: librz/debug/p/common_windows.c (hunk 2 above)
Comment (wargio) [reply librz/debug/p/common_windows.c]: maybe "a second-chance exception occurred." is better.

**Comment by ret2libc** [comment]
Do we have at least one simple test for this thing? Would it be too hard to write one?

File: librz/debug/p/debug_winkd.c (hunk 6 above)
Comment (GustavoLCR) [reply librz/debug/p/debug_winkd.c]: Implemented in 6ba0b4832a

**Comment by GustavoLCR** [comment]
> Do we have at least one simple test for this thing? Would it be too hard to write one?

Some stuff is covered by the kernel dmp tests, but for the winkd protocol itself, theres no simple way of testing it without a windows vm to talk to

**Comment by XVilka** [comment]
```
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_dts continue backwards
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e dbg.create_new_console=false
ood
db @ main
3dc
dr rflags=0x200 > nul
dr rflags
dr rax
dts+
7ds
dr rflags
dr rax
dcb
dr rflags
dr rax
doc
' bins/pe/ioli/w32/crackme0x00.exe
-- stdout
--- expected
+++ actual
@@ -0,7 +0,7 @@
 0x00000200
 0x00000001
-0x00000216
-0x0000001e
-0x00000200
-0x00000001
+0x00000300
+0x44026c4518cd
+0x00000300
+0x44026c4518cd
-- stderr
Spawned new process with pid 3084, tid = 4312
ERROR: File dbg://C:\\projects\\rizin\\test\\bins/pe/ioli/w32/crackme0x00.exe  reopened in read-write mode
hit breakpoint at: 0x401310
Reading 65536 byte(s) from 0x00010000...
Reading 32768 byte(s) from 0x00098000...
Reading 8192 byte(s) from 0x000b0000...
Reading 53248 byte(s) from 0x000d0000...
Reading 32768 byte(s) from 0x001e8000...
Reading 32768 byte(s) from 0x00332000...
Reading 4096 byte(s) from 0x00405000...
Reading 4096 byte(s) from 0x00406000...
Reading 12288 byte(s) from 0x0060d000...
Reading 135168 byte(s) from 0x00680000...
Reading 4096 byte(s) from 0x0097f000...
Reading 36864 byte(s) from 0x00b50000...
Reading 8192 byte(s) from 0x74a1a000...
Reading 4096 byte(s) from 0x75080000...
Reading 12288 byte(s) from 0x75995000...
Reading 16384 byte(s) from 0x75d15000...
Reading 4096 byte(s) from 0x77404000...
Reading 24576 byte(s) from 0x7752e000...
Reading 4096 byte(s) from 0x7ff58000...
Reading 4096 byte(s) from 0x7ff61000...
Reading 4096 byte(s) from 0x7ff80000...
Reading 4096 byte(s) from 0x7ff91000...
Reading 4096 byte(s) from 0x7ff985eea000...
Reading 4096 byte(s) from 0x7ff9865fa000...
Reading 4096 byte(s) from 0x7ff98740f000...
Reading 32768 byte(s) from 0x7ff987412000...
ERROR: invalid register, unable to trace register state
ERROR: invalid register, unable to trace register state
Stepping failed!
Step failed
hit breakpoint at: 0x401310 cnum: 0
rz_debug_reg: error writing registers 0 to 4312
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/43734178/job/6931h4i8d5vxnba2#L4659

**Review by XVilka — APPROVED** [review]
LGTM from the code point of view. But let's wait for @PAI5REECHO answer first.

**Pull request merged**